### PR TITLE
Val loss

### PR DESCRIPTION
**Pull request by maharajamihir**
- implemented val loss for all three models
- implemented train val test split for coinrun data generation
- implemented full frame validation i.e. the entire last frame of a sequence is generated (like during sampling) and the metrics for this frame are calculated

### CODE REVIEW

File: input_pipeline/pngs_to_array_records.py [hunk 1]
```diff
@@ -70,9 +73,30 @@ def preprocess_pngs(input_dir, original_fps, target_fps, chunk_size, target_widt
...
+    results_batch, chunks, file_idx = save_chunks(chunks, file_idx, chunks_per_file, output_path) 
+    results.extend(results_batch)
+    print(f"Done processing files. Saved to {output_path}")
+    return results
+
 def main():
     args = tyro.cli(Args)
     print(f"Output path: {args.output_path}")
+    total_ratio = args.train_ratio + args.val_ratio + args.test_ratio
+    assert np.isclose(total_ratio, 1.0), "Ratios must sum to 1.0"
```
Comment (emergenz) [inline input_pipeline/pngs_to_array_records.py:99]: nice!

File: input_pipeline/generate_coinrun_dataset.py [hunk 2]
```diff
@@ -30,57 +33,83 @@ class Args:
...
+
+
+            ep_metadata, chunks, file_idx = save_chunks(chunks, file_idx, args.chunks_per_file, output_dir_split)
+            episode_metadata.extend(ep_metadata)
+
+            print(f"Episode {episode_idx} completed, length: {step_t + 1}.")
+            episode_idx += 1
+        else:
+            print(f"Episode too short ({step_t + 1}), resampling...")
+    ep_metadata, chunks, file_idx = save_chunks(chunks, file_idx, args.chunks_per_file, output_dir_split)
```
Comment (maharajamihir) [inline input_pipeline/generate_coinrun_dataset.py:80]:
```suggestion
    ep_metadata, _, _ = save_chunks(chunks, file_idx, args.chunks_per_file, output_dir_split)
```

**Review by emergenz — CHANGES_REQUESTED** [review]

File: models/dynamics.py [hunk 3]
```diff
@@ -94,8 +95,14 @@ def __call__(
             vid_embed_BTNM = jnp.where(
                 jnp.expand_dims(mask, -1), self.mask_token.value, vid_embed_BTNM
             )
+        elif pred_full_frame:
+            mask = jnp.zeros_like(video_tokens_BTN)
+            mask = mask.at[: -1].set(True)
```
Comment (emergenz) [inline models/dynamics.py:100]: This is the batch dimension. Is this intentional?

**Review by emergenz — CHANGES_REQUESTED** [review]

File: train_lam.py [hunk 4]
```diff
@@ -305,21 +330,26 @@ def main(args: Args) -> None:
...
+        val_iterator = build_dataloader(args, args.val_data_dir)
 
     # --- Restore checkpoint ---
-    step, optimizer, grain_iterator = restore_checkpoint_if_needed(
-        args, checkpoint_manager, optimizer, grain_iterator
-    )
-
+    if args.val_data_dir:
+        step, optimizer, train_iterator, val_iterator = restore_checkpoint_if_needed(
+            args, checkpoint_manager, optimizer, train_iterator, val_iterator
```
Comment (emergenz) [inline train_lam.py:340]: variable unbound if `not args.val_data_dir`

**Review by emergenz — CHANGES_REQUESTED** [review]

**Review by emergenz — CHANGES_REQUESTED** [review]
The call method of network classes should solely implement the model forward pass. In particular, we should not have scans over 'forward passes' in the call method.

File: train_lam.py [hunk 5]
```diff
@@ -379,26 +411,64 @@ def loss_fn(
...
+        (loss, (recon, _, metrics)) = lam_loss_fn(lam, inputs, training=False)
+        return loss, recon, metrics
+
+    def calculate_validation_metrics(val_dataloader, lam):
+        step = 0
+        loss_per_step = []
+        metrics_per_step = []
+        inputs = None
+        recon = None
+        for videos in val_dataloader:
```
Comment (emergenz) [inline train_lam.py:426]: Since val_dataloader is not returned, this will sample the same videos on every function call.

**Review by emergenz — CHANGES_REQUESTED** [review]

File: train_lam.py [hunk 6]
```diff
@@ -413,23 +483,43 @@ def loss_fn(
...
+                    wandb.log(log_dict)
                 if step % args.log_image_interval == 0:
                     gt_seq = inputs["videos"][0, 1:].astype(jnp.float32) / 255.0
                     recon_seq = recon[0].clip(0, 1)
                     comparison_seq = jnp.concatenate((gt_seq, recon_seq), axis=1)
                     comparison_seq = einops.rearrange(
                         comparison_seq * 255, "t h w c -> h (t w) c"
                     )
+                    if val_results:
+                        val_results["gt_seq_val"] = val_results["gt_batch"]["videos"][0, 1:].astype(jnp.float32) / 255.0
```
Comment (emergenz) [inline train_lam.py:517]: Here, I don't think we should mutate in-place since this could lead to edge cases, e.g. when this code is executed twice consecutively without the `val_results` dict being reinitialized.

**Review by emergenz — CHANGES_REQUESTED** [review]

**Comment by emergenz** [comment]
https://github.com/p-doom/jasmine/pull/190#discussion_r2353331917

**Comment by emergenz** [comment]
https://github.com/p-doom/jasmine/pull/190#discussion_r2353340456

**Comment by emergenz** [comment]
Please also address the two comments mentioned above.

File: input_pipeline/pngs_to_array_records.py [hunk 7]
```diff
@@ -88,40 +112,55 @@ def main():
     else: 
         episodes = directories
 
-    num_processes = mp.cpu_count()
-    print(f"Number of processes: {num_processes}")
-    pool_args = [
-        (
+    n_total = sum([len(os.listdir(episode)) for episode in episodes])
+    n_train = int(n_total * args.train_ratio)
+    n_val = int(n_total * args.val_ratio)
```
Comment (maharajamihir) [inline input_pipeline/pngs_to_array_records.py:119]:
```suggestion
    n_val = round(n_total * args.val_ratio)
```
